lib/srv/db/spanner/grpcserver.go

```diff
@@ -0,0 +1,454 @@
+	// a service account in a different project?
+	dbUser := databaseUserToGCPServiceAccount(e.sessionCtx)
+	ts, err := e.Auth.GetSpannerTokenSource(ctx, e.sessionCtx.WithUser(dbUser))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// The context passed to DialPool controls dialing, but more importantly it
+	// controls GCP oauth2 credential refreshing.
+	// We do not want credential refreshing to stop working after the first
+	// oauth token expires (by default after one hour).
+	// Therefore use the client connection context rather than the RPC context.
+	cc, err := gtransport.DialPool(e.clientConnContext,
+		append(spannerapi.DefaultClientOptions(),
+			option.WithGRPCDialOption(grpc.WithTransportCredentials(credentials.NewTLS(tlsCfg))),
+			option.WithTokenSource(ts),
+			option.WithEndpoint(e.sessionCtx.Database.GetURI()),
+			// pool size seems to be adjusted to number of gRPC channels, which is
+			// 4 by default in Google's code and seemingly other clients as well,
+			// but with Teleport in the middle each downstream client connection
+			// will be handled by a separate engine.
+			// Therefore, avoid amplifying the number of connections dialed to
+			// GCP and just use a pool size of 1.
+			option.WithGRPCConnectionPool(1))...,
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	e.gcloudClient = cc
+	return spannerpb.NewSpannerClient(e.gcloudClient), trace.Wrap(err)
+}
+
+// authorizeRPCRequest checks authorization for an RPC and handles audit
+// logging events based on the result of that check.
+//
+// Spanner IDs look like this:
+// project ID: projects/<project>
+// instance ID: <project ID>/instances/<instance>
+// database ID: <instance ID>/databases/<database>
+// session ID: <database ID>/sessions/<session>
+//
+// All Spanner RPCs either include a database ID or a session ID.
+// As outlined above, a session ID is prefixed by a database ID, and a database
+// ID provides the target GCP project, instance, and database name.
+// This function therefore expects a target ID that looks like
+// `projects/<project>/instances/<instance>/databases/<database>[/sessions/<session]`
+// and will extract the project, instance, and database name to check access.
+func (e *Engine) authorizeRPCRequest(ctx context.Context, targetID, procedure string, req any) error {
+	var (
+		dbName        string
+		authzErr      error
+		checkedAccess bool
+	)
+	e.startSessionOnce.Do(func() {
+		// this signals to other RPCs that the session has started
+		defer close(e.waitForSessionStart)
+		dbName, authzErr = e.checkAccess(ctx, targetID)
+		checkedAccess = true
+		sessionCtx := e.sessionCtx
+		if dbName != "" {
+			sessionCtx = sessionCtx.WithDatabase(dbName)
+		}
+		e.Audit.OnSessionStart(e.Context, sessionCtx, authzErr)
+		e.sessionStartErr = authzErr
+		if authzErr != nil {
+			e.cancelContext()
+		}
+	})
+	// wait for the session to be started
+	<-e.waitForSessionStart
```

Comment:
is this wait necessary?

Reply:
Even though there is only one client connection, there can be concurrent RPCs sent over that connection.
I wanted to keep the audit logging consistent with our other protocols, so I did not want a failed session start to be followed by rejected RPC audit events.
From the perspective of the goroutine that executes the `Once` above, it's not necessary - but there may be other RPC handler goroutines waiting here.

Reply:
I believe that `Do` guarantees when it returns, the callback function has completed. am I missing something?

Reply:
TIL! I did not realize that Once blocked until the first caller returned.

I can now just change `waitForSessionStart` from a chan to a simple `bool` and get rid of the select at the end of HandleConnection too, nice!

Reply:
done [958597d](https://github.com/gravitational/teleport/pull/40859/commits/958597df0432364da130db26550b6fe83bb1cdab)